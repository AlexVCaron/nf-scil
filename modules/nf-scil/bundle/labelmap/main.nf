
process BUNDLE_LABEL_AND_DISTANCE_MAPS {
    tag "$meta.id"
    label 'process_single'

    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
        'https://scil.usherbrooke.ca/containers/scilus_1.6.0.sif':
        'scilus/scilus:1.6.0' }"

    input:
        tuple val(meta), path(bundles), path(centroids)

    output:
    tuple val(meta), path("*_labels.nii.gz")    , emit: label
    tuple val(meta), path("*_labels.trk")       , emit: label_trk
    tuple val(meta), path("*_distances.nii.gz") , emit: distances
    tuple val(meta), path("*_distances.trk")    , emit: distances_trk
    path "versions.yml"                         , emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    def prefix = task.ext.prefix ?: "${meta.id}"
    def nb_points = task.ext.nb_points ?: ''

    String bundles_list = bundles.join(", ").replace(',', '')
    """
    for bundle in $bundles_list;
        do if [[ \$bundle == *"__"* ]]; then
            pos=\$((\$(echo \$bundle | grep -b -o __ | cut -d: -f1)+2))
            bname=\${bundle:\$pos}
            bname=\$(basename \$bname .trk)
        else
            bname=\$(basename \$bundle .trk)
        fi
        bname=\${bname/_ic/}

        centroid=${prefix}__\${bname}_centroid_${nb_points}.trk
        if [[ -f \${centroid} ]]; then
            scil_bundle_label_map.py \$bundle \${centroid} tmp_out -f

            mv tmp_out/labels_map.nii.gz ${prefix}__\${bname}_labels.nii.gz
            mv tmp_out/distance_map.nii.gz ${prefix}__\${bname}_distances.nii.gz

            mv tmp_out/labels.trk ${prefix}__\${bname}_labels.trk
            mv tmp_out/distance.trk ${prefix}__\${bname}_distances.trk
        fi

    done

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        : \$(echo \$(samtools --version 2>&1) | sed 's/^.*samtools //; s/Using.*\$//' ))
    END_VERSIONS
    """

    stub:
    def args = task.ext.args ?: ''
    def prefix = task.ext.prefix ?: "${meta.id}"
    """
    scil_bundle_label_map.py -h
    touch ${prefix}__bundlename_labels.nii.gz
    touch ${prefix}__bundlename_labels.trk
    touch ${prefix}__bundlename_distances.nii.gz
    touch ${prefix}__bundlename_distances.trk

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        : \$(echo \$(samtools --version 2>&1) | sed 's/^.*samtools //; s/Using.*\$//' ))
    END_VERSIONS
    """
}
