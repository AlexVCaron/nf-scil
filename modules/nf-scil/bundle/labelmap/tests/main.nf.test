nextflow_process {

    name "Test Process BUNDLE_LABELMAP"
    script "../main.nf"
    process "BUNDLE_LABELMAP"

    tag "modules"
    tag "modules_nfcore"
    tag "bundle"
    tag "bundle/labelmap"

    tag "subworkflows/load_test_data"

    setup {
        run("fetch_bundleseg_atlas", alias: "LOAD_DATA") {
            script "../../../../../subworkflows/bundle_seg/main.nf"
            process {
                """
                input[0] = "https://zenodo.org/records/10103446/files/atlas.zip?download=1"
                input[1] = "https://zenodo.org/records/10103446/files/config.zip?download=1"
                input[2] = "./"
                """
            }
        }

    }

    test("bundle - labelmap") {

        when {
            process {
                """
                input[0] = file(params.test_data['sarscov2']['illumina']['test_single_end_bam'], checkIfExists: true)
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }

    }

}
