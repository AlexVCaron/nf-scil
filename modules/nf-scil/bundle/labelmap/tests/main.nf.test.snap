{
    "bundle - labelmap": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__IFGWM_labels.nii.gz:md5,5642be1fda227b865de52000f9117123"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__IFGWM_labels.trk:md5,2c7dd1f33c9201a0ce8d86e13c4639d6"
                    ]
                ],
                "2": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__IFGWM_distances.nii.gz:md5,72fdc04b0abdf55de4306e03cc4d162a"
                    ]
                ],
                "3": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__IFGWM_distances.trk:md5,8cd4154a3c4349c2c5141597ce2216bf"
                    ]
                ],
                "4": [
                    "versions.yml:md5,c126a7422f280ed6bd664cc0f6636931"
                ],
                "distances": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__IFGWM_distances.nii.gz:md5,72fdc04b0abdf55de4306e03cc4d162a"
                    ]
                ],
                "distances_trk": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__IFGWM_distances.trk:md5,8cd4154a3c4349c2c5141597ce2216bf"
                    ]
                ],
                "labels": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__IFGWM_labels.nii.gz:md5,5642be1fda227b865de52000f9117123"
                    ]
                ],
                "labels_trk": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__IFGWM_labels.trk:md5,2c7dd1f33c9201a0ce8d86e13c4639d6"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,c126a7422f280ed6bd664cc0f6636931"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "24.04.1"
        },
        "timestamp": "2024-05-22T12:15:08.854169"
    }
}